fcrepo-http-api/src/main/java/org/fcrepo/http/api/DefaultPathLockManager.java

```diff
@@ -0,0 +1,353 @@
+        @Override
+        public void release() {
+            synchronized (DefaultPathLockManager.this) {
+                for (final ActivePath.PathScopedLock lock : locks) {
+                    lock.unlock();
+                    lock.getPath().threads.remove(Thread.currentThread());
+                    if (lock.getPath().threads.isEmpty()) {
+                        activePaths.remove(lock.getPath().path);
+                    }
+                }
+                if (deletePath != null) {
+                    LOGGER.trace("Thread {} releasing delete lock on path {}.",
+                            Thread.currentThread().getId(), deletePath);
+                    activeDeletePaths.remove(deletePath);
+                }
+                LOGGER.trace("Thread {} released locks.", Thread.currentThread().getId());
+                DefaultPathLockManager.this.notify();
+            }
+        }
+
+    }
+
+    /*
+     * This is the only method that adds paths to the pool of
+     * currently active paths that can be locked.
+     */
+    private synchronized ActivePath getActivePath(final String path) {
+        ActivePath activePath = activePaths.get(path);
+        if (activePath == null) {
+            activePath = new ActivePath(path);
+            activePaths.put(path, activePath);
+        }
+        return activePath;
+    }
+
+    private boolean isOrIsDescendantOf(final String possibleDescendant, final String path) {
+        return path.equals(possibleDescendant) || possibleDescendant.startsWith(path + "/");
+    }
+
+    private String getParentPath(final String path) {
+        if (path.indexOf('/') == -1) {
+            return null;
+        }
+        return path.substring(0, path.lastIndexOf('/'));
+    }
+
+    @VisibleForTesting
+    static String normalizePath(final String path) {
+        if (path.endsWith("/")) {
+            return path.substring(0, path.length() - 1);
+        } else {
+            return path;
+        }
+    }
+
+    @Override
+    public AcquiredLock lockForRead(final String path, final Session session, final NodeService nodeService)
+            throws InterruptedException {
+        final List<ActivePath.PathScopedLock> locks = new ArrayList<>();
+
+        synchronized (this) {
+            locks.add(getActivePath(normalizePath(path)).getReadLock());
+        }
+
+        return new AcquiredMultiPathLock(locks);
+    }
+
+    @Override
+    public AcquiredLock lockForWrite(final String path, final Session session, final NodeService nodeService)
+            throws InterruptedException {
+        final List<ActivePath.PathScopedLock> locks = new ArrayList<>();
+
+        synchronized (this) {
+            for (String currentPath = normalizePath(path) ;
+                    currentPath == null || currentPath.length() > 0;
+                    currentPath = getParentPath(currentPath)) {
+                if (currentPath == null || (currentPath != path && nodeService.exists(session, currentPath))) {
```

Comment:
Are you intentionally using `!=` instead of `! currentPath.equals(path)` for comparing these strings?